packages/google_maps_flutter/google_maps_flutter/ios/Classes/GoogleMapTileOverlayController.m

```diff
@@ -0,0 +1,201 @@
+                }
+                if ([result isKindOfClass:[FlutterMethodNotImplemented class]]) {
+                  NSLog(@"Can't get tile: notImplemented");
+                }
+                tileImage = kGMSTileLayerNoTile;
+              }
+
+              [receiver receiveTileWithX:x y:y zoom:zoom image:tileImage];
+            }];
+}
+@end
+
+static BOOL ToBool(NSNumber* data) { return [FLTGoogleMapJsonConversions toBool:data]; }
+
+static float ToFloat(NSNumber* data) { return [FLTGoogleMapJsonConversions toFloat:data]; }
+
+static int ToInt(NSNumber* data) { return [FLTGoogleMapJsonConversions toInt:data]; }
+
+static void InterpretTileOverlayOptions(NSDictionary* data,
+                                        id<FLTGoogleMapTileOverlayOptionsSink> sink,
+                                        NSObject<FlutterPluginRegistrar>* registrar) {
+  NSNumber* visible = data[@"visible"];
+  if (visible != nil) {
+    [sink setVisible:ToBool(visible)];
+  }
+
+  NSNumber* transparency = data[@"transparency"];
+  if (transparency != nil) {
+    [sink setTransparency:ToFloat(transparency)];
+  }
+
+  NSNumber* zIndex = data[@"zIndex"];
+  if (zIndex != nil) {
+    [sink setZIndex:ToInt(zIndex)];
+  }
+
+  NSNumber* fadeIn = data[@"fadeIn"];
+  if (fadeIn != nil) {
+    [sink setFadeIn:ToBool(fadeIn)];
+  }
+
+  NSNumber* tileSize = data[@"tileSize"];
+  if (tileSize != nil) {
+    [sink setTileSize:[tileSize integerValue]];
+  }
+}
+
+@implementation FLTTileOverlaysController {
+  NSMutableDictionary* _tileOverlayIdToController;
+  FlutterMethodChannel* _methodChannel;
+  NSObject<FlutterPluginRegistrar>* _registrar;
+  GMSMapView* _mapView;
+}
+- (instancetype)init:(FlutterMethodChannel*)methodChannel
+             mapView:(GMSMapView*)mapView
+           registrar:(NSObject<FlutterPluginRegistrar>*)registrar {
+  self = [super init];
+  if (self) {
+    _methodChannel = methodChannel;
+    _mapView = mapView;
+    _tileOverlayIdToController = [NSMutableDictionary dictionaryWithCapacity:1];
+    _registrar = registrar;
+  }
+  return self;
+}
+
+- (void)addTileOverlays:(NSArray*)tileOverlaysToAdd {
+  for (NSDictionary* tileOverlay in tileOverlaysToAdd) {
+    NSString* tileOverlayId = [FLTTileOverlaysController getTileOverlayId:tileOverlay];
+    FLTTileProviderController* tileProvider =
+        [[FLTTileProviderController alloc] init:_methodChannel tileOverlayId:tileOverlayId];
+    FLTGoogleMapTileOverlayController* controller =
+        [[FLTGoogleMapTileOverlayController alloc] initWithTileLayer:tileProvider mapView:_mapView];
+    InterpretTileOverlayOptions(tileOverlay, controller, _registrar);
+    _tileOverlayIdToController[tileOverlayId] = controller;
+  }
+}
+
+- (void)changeTileOverlays:(NSArray*)tileOverlaysToChange {
+  for (NSDictionary* tileOverlay in tileOverlaysToChange) {
+    NSString* tileOverlayId = [FLTTileOverlaysController getTileOverlayId:tileOverlay];
```

Comment:
`tileOverlayId` can be nil, the below statement is unsafe.

Reply:
Sorry but I don't see how `tileOverlayId` could be nil, since the string property comes from dart `TileOverlayId.value`, which asserts that `value != null`.